lms/services/digest/_digest_assistant.py

```diff
@@ -0,0 +1,126 @@
+from typing import Iterable, List
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import aggregate_order_by
+
+from lms.models import AssignmentGrouping, AssignmentMembership, Grouping, LTIRole, User
+from lms.models.lti_role import RoleScope, RoleType
+from lms.services.digest._models import HCourse, HUser
+
+
+class DigestAssistant:
+    """A close collaborator of the digest service for dealing with the DB."""
+
+    def __init__(self, db_session):
+        self._db = db_session
+
+    def get_h_users(self, h_userids: Iterable[str]) -> List[HUser]:
+        """
+        Get H users with a username and email.
+
+        This will pick the most recently non-null emails and display names for
+        the user.
+
+        :param h_userids: List of H userids (e.g. 'acct:user@lms.example.com')
+        """
+
+        query = (
+            # The column names here must match the HUser attributes
+            sa.select(
+                User.h_userid,
+                # Aggregate all the emails in order of most recently updated
+                # then filter out any which are NULL and pick the first.
+                sa.func.array_agg(aggregate_order_by(User.email, User.updated.desc()))
+                .filter(User.email.isnot(None))[1]
+                .label("email"),
+                # Similar for the display name
+                sa.func.array_agg(
+                    aggregate_order_by(User.display_name, User.updated.desc())
+                )
+                .filter(User.display_name.isnot(None))[1]
+                .label("display_name"),
+            )
+            .where(User.h_userid.in_(h_userids))
+            .group_by(User.h_userid)
+        )
+
+        return self._query_to_dataclass(query, HUser)
+
+    def get_h_courses(self, authority_provided_ids: Iterable[str]) -> List[HCourse]:
+        """
+        Get courses along with any child sections and teachers.
+
+        This will look for any courses which have the given authority provided
+        id, or have children with the given authority provided id.
+
+        This will return a list of course object which includes other authority
+        provided ids which might link to this group (from sections etc.) and
+        a list of H userids for instructors in the course.
+
+        :param authority_provided_ids: A list of authority provided ids to
+            start the search.
+        """
+
```

Comment:
This is a belter and no mistake.

I don't think SQLAlchemy makes this any easier on us here. I find the raw SQL considerably easier to read. However it is convenient in a few places, as we can directly refer to things we would have to use binds for otherwise.

In any case, it does show that we can do fairly complicated SQL in SQLAlchemy. It does require me to search the internet for nearly every part of it though, which I dislike. It's often pretty hard to know where to start with something like:

```(ARRAY_AGG(col ORDER BY foo) FILTER (WHERE thing)```

Half of that is in the main library, some is Postgres specific, and there's no rhyme or reason to the syntax required.